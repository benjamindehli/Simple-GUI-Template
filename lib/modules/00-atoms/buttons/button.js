"use strict";
exports.__esModule = true;
var Style = require("../../../../src/modules/00-atoms/buttons/button.scss");
var Button = (function () {
    function Button(button) {
        this.id = "";
        this.link = "#";
        this.icon = "";
        this.content = "";
        this.type = "flat";
        this.theme = "default";
        this.submit = false;
        this.id = button.id;
        if (button.id !== undefined)
            this.id = button.id;
        if (button.link !== undefined)
            this.link = button.link;
        if (button.iconClass !== undefined)
            this.icon = "<span class='" + Style.icon + " " + button.iconClass + "'></span>";
        if (button.content !== undefined)
            this.content = button.content;
        if (button.title !== undefined)
            this.title = button.title;
        if (button.type !== undefined)
            this.type = button.type;
        if (button.theme !== undefined)
            this.theme = button.theme;
        if (button.submit !== undefined)
            this.submit = button.submit;
        if (button.attributes !== undefined)
            this.attributes = button.attributes;
        if (button.formWrapper !== undefined) {
            this.formWrapper = {
                formAction: button.formWrapper.formAction !== undefined ? button.formWrapper.formAction : '',
                formMethod: button.formWrapper.formMethod !== undefined ? button.formWrapper.formMethod : ''
            };
        }
    }
    Button.prototype.getThemeClass = function (theme) {
        var themeClass = Style.buttonThemeDefault;
        if (theme == "primary")
            themeClass = Style.buttonThemePrimary;
        if (theme == "info")
            themeClass = Style.buttonThemeInfo;
        if (theme == "success")
            themeClass = Style.buttonThemeSuccess;
        if (theme == "warning")
            themeClass = Style.buttonThemeWarning;
        if (theme == "danger")
            themeClass = Style.buttonThemeDanger;
        return themeClass;
    };
    Button.prototype.getTypeClass = function (type) {
        var typeClass = Style.buttonTypeFlat;
        if (type == "raised")
            typeClass = Style.buttonTypeRaised;
        if (type == "minimal")
            typeClass = Style.buttonTypeMinimal;
        return typeClass;
    };
    Button.prototype.getHtmlAttributes = function (attributes) {
        var htmlAttributes = '';
        attributes.forEach(function (attribute, index) {
            htmlAttributes += attribute;
            if (index < attributes.length) {
                htmlAttributes += ' ';
            }
        });
        return htmlAttributes;
    };
    Button.prototype.addFormWrapper = function (moduleElement) {
        return "<form action='" + this.formWrapper.formAction + "' method='" + this.formWrapper.formMethod + "' class='" + Style.formWrapper + "'>" + moduleElement + "</form>";
    };
    Button.prototype.createModuleElement = function () {
        var themeClass = this.getThemeClass(this.theme);
        var typeClass = (this.type !== undefined) ? this.getTypeClass(this.type) : Style.buttonTypeFlat;
        var titleAttribute = (this.title !== undefined) ? "title='" + this.title + "'" : '';
        var isSubmitButton = this.submit !== undefined && this.submit === true;
        var htmlAtributes = this.attributes !== undefined && this.attributes.length ? this.getHtmlAttributes(this.attributes) : '';
        var moduleElement = '';
        if (isSubmitButton) {
            moduleElement = "<button type='submit' id='" + this.id + "' " + htmlAtributes + " class='" + Style.button + " " + typeClass + " " + themeClass + "'>" + this.icon + " " + this.content + "</button>";
        }
        else {
            moduleElement = "<a id='" + this.id + "' " + titleAttribute + " href='" + this.link + "' " + htmlAtributes + " class='" + Style.button + " " + typeClass + " " + themeClass + "'>" + this.icon + " " + this.content + "</a>";
        }
        if (this.formWrapper !== undefined) {
            moduleElement = this.addFormWrapper(moduleElement);
        }
        return moduleElement;
    };
    return Button;
}());
exports.Button = Button;
function getModule(button) {
    return new Button(button).createModuleElement();
}
exports.getModule = getModule;
