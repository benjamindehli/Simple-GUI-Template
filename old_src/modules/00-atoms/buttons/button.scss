@import "../../../../src/style/abstracts/all";
:local(.button) {
	cursor: pointer;
	text-align: center;
	min-width: 42px;
	height: 34px;
	border: 1px solid rgba(0, 0, 0, .07);
	font-size: 14px;
	font-family: $normal-font;
	font-weight: bold;
	color: $primary-color;
	line-height: 32px;
	display: inline-block;
	padding: 0 6px;
	text-transform: uppercase;
	@include border-radius(2px);
	&[disabled] {
		cursor: not-allowed;
	}
}

:local(.formWrapper){
	display: inline-block;
	vertical-align: bottom;
}

:local(.icon){
	font-size: 14pt!important;
	vertical-align: middle;
}

:local(.buttonTypeFlat) {
	@include transition(all .2s);

	&:not([disabled]):hover {
		background-color: bright-gray($dark-color);
		color: #FFFFFF;
	}
	&:not([disabled]):active {
		background-color: bright-color($primary-color);
	}
	&[disabled] {
		color: rgba(0, 0, 0, 0.26);
		background-color: transparent;
	}
}

:local(.buttonTypeRaised) {
	&:not([disabled]) {
		@include raised(2);
		@include transition(box-shadow .4s cubic-bezier(.25, .8, .25, 1), background-color .4s cubic-bezier(.25, .8, .25, 1));
		&:hover{
			color: #FFFFFF;
		}
	}
	&:not([disabled]):active {
		@include raised(3);
		background-color: bright-color($primary-color);
	}
	&[disabled] {
		color: rgba(0, 0, 0, 0.26);
	}
}

:local(.buttonTypeMinimal)[disabled] {
	background: none;
	border: none;
	color: rgba(0, 0, 0, 0.26);
}

:local(.buttonTypeMinimal):not([disabled]) {
	background: none;
	border: none;
	color: $primary-color;
	&:local(.buttonThemeDefault),
	&:local(.buttonThemeDefault):visited,
	&:local(.buttonThemePrimary),
	&:local(.buttonThemePrimary):visited,
	&:local(.buttonThemeInfo),
	&:local(.buttonThemeInfo):visited,
	&:local(.buttonThemeSuccess),
	&:local(.buttonThemeSuccess):visited,
	&:local(.buttonThemeWarning),
	&:local(.buttonThemeWarning):visited,
	&:local(.buttonThemeDanger),
	&:local(.buttonThemeDanger):visited{
		background: none;
		border: none;
		&:hover {
			background-color: bright-gray($dark-color);
		}
	}
	&:local(.buttonThemeDefault),
	&:local(.buttonThemeDefault):visited {
		color: $default-text-color;
		&:active {
			background-color: bright-gray-active($dark-color);
		}
	}
	&:local(.buttonThemePrimary),
	&:local(.buttonThemePrimary):visited {
		color: $primary-color;
		&:active {
			background-color: bright-color($primary-color);
		}
	}
	&:local(.buttonThemeInfo),
	&:local(.buttonThemeInfo):visited {
		color: $info-color;
		&:active {
			background-color: bright-color($info-color);
		}
	}
	&:local(.buttonThemeSuccess),
	&:local(.buttonThemeSuccess):visited {
		color: $success-color;
		&:active {
			background-color: bright-color($success-color);
		}
	}
	&:local(.buttonThemeWarning),
	&:local(.buttonThemeWarning):visited {
		color: $warning-color;
		&:active {
			background-color: bright-color($warning-color);
		}
	}
	&:local(.buttonThemeDanger),
	&:local(.buttonThemeDanger):visited {
		color: $danger-color;
		&:active {
			background-color: bright-color($danger-color);
		}
	}
}

:local(.buttonThemeDefault):not([disabled]),
:local(.buttonThemeDefault):not([disabled]):visited {
	background-color: $default-color;
	color: #333;
	border: 1px solid border-color($default-color);
	@include transition(all .2s);
	&:hover {
		background-color: $default-hover-color;
		color: #333;
	}
	&:active {
		background-color: $default-active-color;
		color: #333;
	}
	&:local(.buttonTypeFlat) {
		&:active {
			background-color: $default-active-color;
			color: #333;
		}
	}
}

:local(.buttonThemePrimary):not([disabled]),
:local(.buttonThemePrimary):not([disabled]):visited {
	background-color: $primary-color;
	color: #FFFFFF;
	@include transition(all .2s);
	&:hover {
		background-color: hover-color($primary-color);
	}
	&:active {
		background-color: active-color($primary-color);
	}
	&:local(.buttonTypeFlat) {
		&:active {
			background-color: active-color($primary-color);
		}
	}
}

:local(.buttonThemeInfo):not([disabled]),
:local(.buttonThemeInfo):not([disabled]):visited {
	background-color: $info-color;
	color: #FFFFFF;
	@include transition(all .2s);
	&:hover {
		background-color: hover-color($info-color);
	}
	&:active {
		background-color: active-color($info-color);
	}
	&:local(.buttonTypeFlat) {
		&:active {
			background-color: active-color($info-color);
		}
	}
}

:local(.buttonThemeSuccess):not([disabled]),
:local(.buttonThemeSuccess):not([disabled]):visited {
	background-color: $success-color;
	color: #FFFFFF;
	@include transition(all .2s);
	&:hover {
		background-color: hover-color($success-color);
	}
	&:active {
		background-color: active-color($success-color);
	}
	&:local(.buttonTypeFlat) {
		&:active {
			background-color: active-color($success-color);
		}
	}
}

:local(.buttonThemeWarning):not([disabled]),
:local(.buttonThemeWarning):not([disabled]):visited {
	background-color: $warning-color;
	color: #FFFFFF;
	@include transition(all .2s);
	&:hover {
		background-color: hover-color($warning-color);
	}
	&:active {
		background-color: active-color($warning-color);
	}
	&:local(.buttonTypeFlat) {
		&:active {
			background-color: active-color($warning-color);
		}
	}
}

:local(.buttonThemeDanger):not([disabled]),
:local(.buttonThemeDanger):not([disabled]):visited {
	background-color: $danger-color;
	color: #FFFFFF;
	@include transition(all .2s);
	&:hover {
		background-color: hover-color($danger-color);
	}
	&:active {
		background-color: active-color($danger-color);
	}
	&:local(.buttonTypeFlat) {
		&:active {
			background-color: active-color($danger-color);
		}
	}
}