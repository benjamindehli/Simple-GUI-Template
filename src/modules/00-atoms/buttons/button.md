# Button
```json
{
	"id": "ID for button",
	"type": "Button type [flat, raised]",
	"theme": "Button theme [default, primary, info, success, warning, danger]",
	"link": "Href attribute value for button element",
	"iconClass": "Button icon before content",
	"content": "Content inside button",
	"title": "Title attribute value for button element",
	"submit" "Submit button for forms [true/false]"
}
```
