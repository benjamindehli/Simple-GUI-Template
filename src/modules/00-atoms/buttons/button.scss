@import "../../../../src/style/abstracts/all";
:local(.button) {
	cursor: pointer;
	text-align: center;
	min-width: 42px;
	height: 34px;
	border: 1px solid rgba(0, 0, 0, .07);
	font-size: 14px;
	font-weight: bold;
	color: $primary-color;
	line-height: 32px;
	display: inline-block;
	margin: 0 8px;
	padding: 0 6px;
	text-transform: uppercase;
	@include border-radius(2px);
}

:local(.icon){
	font-size: 14pt!important;
	vertical-align: middle;
}

:local(.buttonTypeFlat) {
	@include transition(all .2s .1s);

	&:not([disabled]):hover {
		background-color: bright-gray($dark-color);
	}
	&:not([disabled]):active {
		background-color: bright-color($primary-color);
	}
	&[disabled] {
		color: rgba(0, 0, 0, 0.26);
		background-color: transparent;
	}
}

:local(.buttonTypeRaised) {
	&:not([disabled]) {
		@include raised(2);
		@include transition(box-shadow .4s cubic-bezier(.25, .8, .25, 1), background-color .4s cubic-bezier(.25, .8, .25, 1));
	}
	&:not([disabled]):active {
		@include raised(3);
		background-color: bright-color($primary-color);
	}
	&[disabled] {
		color: rgba(0, 0, 0, 0.26);
	}
}

:local(.buttonTypeMinimal) {
	background: none;
	border: none;
	color: $primary-color;
	&:local(.buttonThemeDefault),
	&:local(.buttonThemeDefault):visited,
	&:local(.buttonThemePrimary),
	&:local(.buttonThemePrimary):visited,
	&:local(.buttonThemeInfo),
	&:local(.buttonThemeInfo):visited,
	&:local(.buttonThemeSuccess),
	&:local(.buttonThemeSuccess):visited,
	&:local(.buttonThemeWarning),
	&:local(.buttonThemeWarning):visited,
	&:local(.buttonThemeDanger),
	&:local(.buttonThemeDanger):visited{
		background: none;
		border: none;
		&:not([disabled]):hover {
			background-color: bright-gray($dark-color);
		}
	}
	&:local(.buttonThemeDefault),
	&:local(.buttonThemeDefault):visited {
		color: $primary-color;
		&:not([disabled]):active {
			background-color: bright-gray-active($dark-color);
		}
	}
	&:local(.buttonThemePrimary),
	&:local(.buttonThemePrimary):visited {
		color: $primary-color;
		&:not([disabled]):active {
			background-color: bright-color($primary-color);
		}
	}
	&:local(.buttonThemeInfo),
	&:local(.buttonThemeInfo):visited {
		color: $info-color;
		&:not([disabled]):active {
			background-color: bright-color($info-color);
		}
	}
	&:local(.buttonThemeSuccess),
	&:local(.buttonThemeSuccess):visited {
		color: $success-color;
		&:not([disabled]):active {
			background-color: bright-color($success-color);
		}
	}
	&:local(.buttonThemeWarning),
	&:local(.buttonThemeWarning):visited {
		color: $warning-color;
		&:not([disabled]):active {
			background-color: bright-color($warning-color);
		}
	}
	&:local(.buttonThemeDanger),
	&:local(.buttonThemeDanger):visited {
		color: $danger-color;
		&:not([disabled]):active {
			background-color: bright-color($danger-color);
		}
	}
}

:local(.buttonThemeDefault),
:local(.buttonThemeDefault):visited {
	background-color: $default-color;
	color: #333;
	border: 1px solid border-color($default-color);
	@include transition(all .2s .1s);
	&:not([disabled]):hover {
		background-color: $default-hover-color;
		color: #333;
	}
	&:not([disabled]):active {
		background-color: $default-active-color;
		color: #333;
	}
	&:local(.buttonTypeFlat) {
		&:not([disabled]):active {
			background-color: $default-active-color;
			color: #333;
		}
	}
}

:local(.buttonThemePrimary),
:local(.buttonThemePrimary):visited {
	background-color: $primary-color;
	color: #FFFFFF;
	@include transition(all .2s .1s);
	&:not([disabled]):hover {
		background-color: hover-color($primary-color);
	}
	&:not([disabled]):active {
		background-color: active-color($primary-color);
	}
	&:local(.buttonTypeFlat) {
		&:not([disabled]):active {
			background-color: active-color($primary-color);
		}
	}
}

:local(.buttonThemeInfo),
:local(.buttonThemeInfo):visited {
	background-color: $info-color;
	color: #FFFFFF;
	@include transition(all .2s .1s);
	&:not([disabled]):hover {
		background-color: hover-color($info-color);
	}
	&:not([disabled]):active {
		background-color: active-color($info-color);
	}
	&:local(.buttonTypeFlat) {
		&:not([disabled]):active {
			background-color: active-color($info-color);
		}
	}
}

:local(.buttonThemeSuccess),
:local(.buttonThemeSuccess):visited {
	background-color: $success-color;
	color: #FFFFFF;
	@include transition(all .2s .1s);
	&:not([disabled]):hover {
		background-color: hover-color($success-color);
	}
	&:not([disabled]):active {
		background-color: active-color($success-color);
	}
	&:local(.buttonTypeFlat) {
		&:not([disabled]):active {
			background-color: active-color($success-color);
		}
	}
}

:local(.buttonThemeWarning),
:local(.buttonThemeWarning):visited {
	background-color: $warning-color;
	color: #FFFFFF;
	@include transition(all .2s .1s);
	&:not([disabled]):hover {
		background-color: hover-color($warning-color);
	}
	&:not([disabled]):active {
		background-color: active-color($warning-color);
	}
	&:local(.buttonTypeFlat) {
		&:not([disabled]):active {
			background-color: active-color($warning-color);
		}
	}
}

:local(.buttonThemeDanger),
:local(.buttonThemeDanger):visited {
	background-color: $danger-color;
	color: #FFFFFF;
	@include transition(all .2s .1s);
	&:not([disabled]):hover {
		background-color: hover-color($danger-color);
	}
	&:not([disabled]):active {
		background-color: active-color($danger-color);
	}
	&:local(.buttonTypeFlat) {
		&:not([disabled]):active {
			background-color: active-color($danger-color);
		}
	}
}