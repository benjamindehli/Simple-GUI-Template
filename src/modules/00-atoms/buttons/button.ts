const Style = require<any>("../../../../src/modules/00-atoms/buttons/button.scss");

export class Button {
	id: string = "";
	link: string = "#";
	icon: string = "";
	content: string = "";
	title: string;
	type?: string = "flat";
	theme?: string = "default";
	submit?: boolean = false;

	constructor(button: IButton) {
		this.id = button.id;
		if (button.id !== undefined) this.id = button.id;
		if (button.link !== undefined) this.link = button.link;
		if (button.iconClass !== undefined) this.icon = `<span class='${Style.icon} ${button.iconClass}'></span>`;
		if (button.content !== undefined ) this.content = button.content;
		if (button.title !== undefined ) this.title = button.title;
		if (button.type !== undefined) this.type = button.type;
		if (button.theme !== undefined) this.theme = button.theme;
		if (button.submit !== undefined) this.submit = button.submit;
	}

	private getThemeClass(theme: string){
		let themeClass = Style.buttonThemeDefault;
		if (theme == "primary")	themeClass = Style.buttonThemePrimary;
		if (theme == "info") 	themeClass = Style.buttonThemeInfo;
		if (theme == "success")	themeClass = Style.buttonThemeSuccess;
		if (theme == "warning")	themeClass = Style.buttonThemeWarning;
		if (theme == "danger") 	themeClass = Style.buttonThemeDanger;
		return themeClass;
	}

	private getTypeClass(type: string){
		let typeClass = Style.buttonTypeFlat;
		if (type == "raised") typeClass = Style.buttonTypeRaised;
		if (type == "minimal") typeClass = Style.buttonTypeMinimal;
		return typeClass;
	}

	public createModuleElement() {
		let themeClass = this.getThemeClass(this.theme)
		let typeClass = (this.type !== undefined) ? this.getTypeClass(this.type) : Style.buttonTypeFlat;
		let titleAttribute: string = (this.title !== undefined) ? `title='${this.title}'` : '';
		let isSubmitButton:boolean = this.submit !== undefined && this.submit === true;
		let moduleElement:string = '';
		if (isSubmitButton){
			moduleElement = `<button type='submit' id='${this.id}' class='${Style.button} ${typeClass} ${themeClass}'>${this.icon} ${this.content}</button>`;
		}else{
			moduleElement = `<a id='${this.id}' ${titleAttribute} href='${this.link}' class='${Style.button} ${typeClass} ${themeClass}'>${this.icon} ${this.content}</a>`;
		}
		return moduleElement;
	}
}

export interface IButton {
	id?: string; 
	type?: string; 
	theme?: string;
	link ?: string; 
	iconClass?: string; 
	content?: string;
	title?: string;
	submit?: boolean;
}

export function getModule(button: IButton){
	return new Button(button).createModuleElement();
}
