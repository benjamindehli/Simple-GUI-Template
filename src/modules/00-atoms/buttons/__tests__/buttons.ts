import * as Button from '../button';


test('adds 1 + 2 to equal 3', () => {
	expect(1 + 2).toBe(3);
});


it('renders as an anchor with id value "testbutton"', () => {
	const buttonElement = Button.getModule({
		id: 'testbutton'
	});
	expect(buttonElement).toMatchSnapshot();
});

it('renders a button anchor with type value "submit"', () => {
	const buttonElement = Button.getModule({
		submit: true
	});
	expect(buttonElement).toMatchSnapshot();
});
