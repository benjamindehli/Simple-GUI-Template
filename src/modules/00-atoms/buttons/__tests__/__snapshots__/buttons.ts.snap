// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a button anchor with type value "submit" 1`] = `"<button type='submit'   class='undefined undefined undefined'><span class='undefined '></span> </button>"`;

exports[`renders as an anchor with id value "testbutton" 1`] = `"<a id='testbutton'  href='#'  class='undefined undefined undefined'><span class='undefined '></span> </a>"`;
